.pre-commit-config.yaml

```diff
@@ -6,10 +6,15 @@ repos:
   - repo: https://gitlab.com/pycqa/flake8
     rev: 4.0.1
     hooks:
-     - id: flake8
+      - id: flake8
   - repo: https://github.com/python/black.git
     rev: 22.3.0
     hooks:
-        - id: black
-          language_version: python3
-          args: ["--check"]
+      - id: black
+        language_version: python3
+        args: ["--check"]
+  - repo: https://github.com/pycqa/isort
+    rev: 5.10.1
+    hooks:
+      - id: isort
+        name: isort (python)
```

Comment:
This is interesting. You're running `black` in `--check` mode but `isort` in the default "fix" mode. If your precommit hook is going to automatically fix the imports... should we just let black fix everything automatically as well? If not, then it should probably also be `isort --check`.

I'm also not really familiar with the syntax here: but why are you using `language_version` for `black` but not for `isort` (or `flake8` for that matter, since they're all Python tools)??

Reply:
I second Dave's comment.

Also, why are we specifying `name` only for `isort` and not for the other two?

(And, why _don't_ we have to also exclude `web-server/v0.3/demo.py` at line 4 (as we do (twice) in `project.toml`)?)

Reply:
Some of the existing code is there for reasons I don't understand.  I just fixed the indentation.

I have applied the feedback based on what I _think_ is reasonable, but still not sure myself.

Reply:
I'm still kinda wondering if we *should* remove the `--check` from both, since we're effectively requiring that it be done manually before we can merge a PR anyway. Then again, it's good they're consistent and we can always decide to do that later ...

Webb's comment raised a secondary consideration: so, how do explicit settings here interact with the `project.toml`? Are we implicitly gaining the defaults we set there as long as we don't override them here, or does this run in a completely independent environment?

Reply:
Last I checked, `pre-commit` runs the commands only against the files being added or changed.  It is not running them against the whole tree.

I'll experiment with removing the duplicate config here.

Reply:
> Last I checked, `pre-commit` runs the commands only against the files being added or changed. It is not running them against the whole tree.

Good point. So the question would be, *if* someone explicitly edits `demo.py`, would the exclusion configuration prevent it from being checked; and, if so, is that what we want? 👀

Reply:
Just posted an updated comment with the answer.

Reply:
If we remove the `--check`, doesn't that default to "fix it!"...which I don't think we want to have done implicitly.

We might want to consider adding the "show the diff" switch, though.

Reply:
> If we remove the --check, doesn't that default to "fix it!"...which I don't think we want to have done implicitly.

That's certainly the "traditional" answer. On the other hand, CI lint tests will fail and we won't allow a merge unless they pass; so an argument could be made that having both `black .` and `isort .` executed automatically via precommit hook would just remove a tedious manual development step with no real downside.

> We might want to consider adding the "show the diff" switch, though.

Yes; for both `black` and `isort`, we could do `--diff --check` for more verbose output showing what would be changed. Although, again, since we essentially require that you use the tools prior to merging it's unclear whether the extra verbosity is valuable...

Reply:
> with no real downside

Given it's limited purview, I might agree for `isort`; however, in the case of Black...no thank you!  There are occasionally times when I prefer to re-_code_ rather than accept Black's choices of reformatting.  So, while having a pre-_check_ on commits is probably a good thing, having code committed under my name be changed without consulting me is highly undesirable.

> we could do --diff --check for more verbose output showing what would be changed. Although, again, since we essentially require that you use the tools prior to merging it's unclear whether the extra verbosity is valuable...

The reason for adding `--diff` is to give the developer context when the `--check` fails and to allow the developer to decide whether to accept the proffered change.  The fact that we require the check to pass isn't quite the same as requiring the developer to accept the _tool's_ choice of change.